### PR TITLE
Speedup link prediction sampling

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*
If we send train_mask to DGL neighbor sampler, the neighbor sampler will be slow when sampling neighbors for super nodes. This PR add a new sampler that does not send train_mask to DGL sampler but use the train_mask to trim the sampled graph.

One epoch time of training 1 layer RGCN on MAG dataset on 4 g5.48x reduces from 28301s to 19051s (1.5X speedup) and the model performance at the end of first epoch is compatible (mrr 0.284 vs. mrr 0.29)

Together with DGL 1.0.4 (#303) one epoch time of training 2 layer RGCN on MAG dataset on 4 g5.48x reduces from 59916.7s to 24811.3s (2.4X speedup).

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/dataloading/sampler.py [hunk 1]
```diff
@@ -243,3 +246,107 @@ def _generate(self, g, eids, canonical_etype):
...
+        reverse_edge_types_map=None,
+    ):
+        self.mask = mask
+        self.reverse_edge_types_map = reverse_edge_types_map
+        self.original_edge_types_map = {
+            val: key for key, val in reverse_edge_types_map.items()
+        } if reverse_edge_types_map is not None else None
+
+        print(self.reverse_edge_types_map)
+        print(self.original_edge_types_map)
```
Comment (zheng-da) [inline python/graphstorm/dataloading/sampler.py:280]: remove the print.

Comment (classicsong) [reply python/graphstorm/dataloading/sampler.py]: removed

File: python/graphstorm/dataloading/sampler.py [hunk 2]
```diff
@@ -243,3 +246,105 @@ def _generate(self, g, eids, canonical_etype):
...
+            #        print(f"{len(new_eid[etype])} vs {new_block.num_edges(etype)}")
+            #    else:
+            #        print(f"No eid etype {etype}")
+
+            #for etype in new_block.canonical_etypes:
+            #    if new_block.num_edges(etype) > 0:
+            #        new_block.edges[etype].data[EID] = new_eid[etype]
+
+            #seed_nodes = new_block.srcdata[NID]
+            #blocks.insert(0, new_block)
```
Comment (zheng-da) [inline python/graphstorm/dataloading/sampler.py:348]: remove the code.

File: python/graphstorm/dataloading/sampler.py [hunk 3]
```diff
@@ -243,3 +246,105 @@ def _generate(self, g, eids, canonical_etype):
...
+            new_eid = dict(eid)
+            if self.mask is not None:
+                new_edges = {}
+                for etype in frontier.canonical_etypes:
+                    if self.mask in g.edges[etype].data:
+                        # train mask in data
+                        if etype in eid:
+                            mask = g.edges[etype].data[self.mask][eid[etype]].bool()
+                            new_edges[etype] = mask
+                            new_eid[etype] = eid[etype][mask]
```
Comment (zheng-da) [inline python/graphstorm/dataloading/sampler.py:314]: is it possible that there are no edges left after the filtering?

File: python/graphstorm/dataloading/sampler.py (hunk 2 above)
Comment (classicsong) [reply python/graphstorm/dataloading/sampler.py:348]: removed

File: python/graphstorm/dataloading/sampler.py (hunk 3 above)
Comment (classicsong) [reply python/graphstorm/dataloading/sampler.py:314]: In theory yes. But with large enough fanout, e.g., 5 or 10 and small val and test sets e.g.., < 10% 5%, the change is very low.

File: python/graphstorm/dataloading/sampler.py [hunk 4]
```diff
@@ -243,3 +246,89 @@ def _generate(self, g, eids, canonical_etype):
...
+
+class FastMultiLayerNeighborSampler(NeighborSampler):
+    """ Fast MultiLayerNeighborSampler
+
+        If mask is None, it acts the same as dgl.dataloading.MultiLayerNeighborSampler
+
+    Parameters
+    ----------
+    reverse_edge_types_map: dict
+        A map for reverse edge type
```
Comment (zheng-da) [inline python/graphstorm/dataloading/sampler.py:258]: this docstring is incomplete.

File: python/graphstorm/dataloading/sampler.py [hunk 5]
```diff
@@ -243,3 +246,89 @@ def _generate(self, g, eids, canonical_etype):
...
+            prob=prob,
+            mask=None, # Do neighbor sampling with out edge mask
+            replace=replace,
+            prefetch_node_feats=prefetch_node_feats,
+            prefetch_labels=prefetch_labels,
+            prefetch_edge_feats=prefetch_edge_feats,
+            output_device=output_device
+        )
+
+    def sample_blocks(self, g, seed_nodes, exclude_eids=None):
```
Comment (zheng-da) [inline python/graphstorm/dataloading/sampler.py:291]: add docstring.

**Comment by zheng-da** [comment]
integration tests and unit tests?

File: python/graphstorm/dataloading/sampler.py (hunk 4 above)
Comment (classicsong) [reply python/graphstorm/dataloading/sampler.py:258]: added

File: python/graphstorm/dataloading/sampler.py (hunk 5 above)
Comment (classicsong) [reply python/graphstorm/dataloading/sampler.py:330]: added

**Comment by classicsong** [comment]
> integration tests and unit tests?

I add some integration tests.
It is very hard to add unit tests.

File: python/graphstorm/dataloading/sampler.py [hunk 6]
```diff
@@ -243,3 +246,144 @@ def _generate(self, g, eids, canonical_etype):
...
+                        # train mask in data
+                        if etype in eid:
+                            mask = g.edges[etype].data[self.mask][eid[etype]].bool()
+                            new_edges[etype] = mask
+                            new_eid[etype] = eid[etype][mask]
+
+                    elif self.original_edge_types_map is not None and \
+                        self.mask in g.edges[self.original_edge_types_map[etype]].data:
+                        # handle reverse edges
+                        original_etype = self.original_edge_types_map[etype]
```
Comment (zheng-da) [inline python/graphstorm/dataloading/sampler.py:374]: is this naming correctly? is this original etype or reversed etype?

**Review by zheng-da — APPROVED** [review]

File: python/graphstorm/dataloading/sampler.py (hunk 6 above)
Comment (classicsong) [reply python/graphstorm/dataloading/sampler.py:374]: It is the reverse map of reverse_edge_types_map.

**Pull request merged**